core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -60,8 +65,8 @@ public class WorkerDynamicCatalogManager
     private final ReadWriteLock catalogsLock = new ReentrantReadWriteLock();
     private final Lock catalogLoadingLock = catalogsLock.readLock();
     private final Lock catalogRemovingLock = catalogsLock.writeLock();
-    private final ConcurrentMap<CatalogHandle, CatalogConnector> catalogs = new ConcurrentHashMap<>();
-    private final ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor();
+    private final ConcurrentMap<CatalogHandle, ListenableFuture<CatalogConnector>> catalogs = new ConcurrentHashMap<>();
```

Comment:
the desired functionality to synchronize on the key being loaded can be achieved using the `Cache` from Guava, which does that. This may result in less futures in the code => simpler.

Reply:
I don't understand how this could work and why this is easier.

Reply:
when i look at what we're trying to accomplish here, it looks that `ConcurrentMap<CatalogHandle, ListenableFuture<CatalogConnector>>` is equivalent to `Cache<CatalogHandle, CatalogConnector>`, as cache handles futures (or equivalent thereof) internally.